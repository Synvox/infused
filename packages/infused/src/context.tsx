import { createContext, ReactNode, useContext } from "react";
import { InfusedStyleSheet } from "./stylesheet";

const context = createContext<{
  register(sheet: InfusedStyleSheet): void;
  sheets: Set<InfusedStyleSheet>;
  replacedString: string;
}>({ register: () => {}, sheets: new Set(), replacedString: "" });

export function useRegisterInfusedStyleSheet(stylesheet: InfusedStyleSheet) {
  useContext(context).register(stylesheet);
}

export function CSSExtractionProvider({
  sheets,
  children,
  replacedString,
}: {
  sheets: Set<InfusedStyleSheet>;
  children: ReactNode;
  replacedString: string;
}) {
  function register(sheet: InfusedStyleSheet) {
    sheets.add(sheet);
  }

  return (
    <context.Provider value={{ register, sheets, replacedString }}>
      {children}
    </context.Provider>
  );
}

export function Styles() {
  if (typeof document !== "undefined") return null;
  return <>{useContext(context).replacedString}</>;
}
