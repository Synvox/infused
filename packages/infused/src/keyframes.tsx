import { InfusedStyleSheet, WithInfusedStyleSheet } from "./stylesheet";
import { template } from "./template";

export class CSSKeyframes extends WithInfusedStyleSheet {
  id: string;
  constructor(p: {
    id: typeof CSSKeyframes.prototype.id;
    stylesheet: typeof CSSKeyframes.prototype.stylesheet;
  }) {
    super(p.stylesheet);
    this.id = p.id;
  }
}

export const _keyframes = (stylesheet: InfusedStyleSheet) =>
  template(
    stylesheet,
    (id) => new CSSKeyframes({ id, stylesheet }),
    () => ({
      end: (str) => (id: string) => `@keyframes ${id}{${str}}`,
    })
  );
