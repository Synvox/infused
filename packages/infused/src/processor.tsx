export type Processor<T> = {
  next?(string: string, charCode: number): void;
  end(fullString: string): T;
};

export function process<T>(str: string, fn: () => Processor<T>) {
  const length = str.length;
  const process = fn();

  if (process.next)
    for (let i = 0; i < length; i++) {
      process.next(str[i], str.charCodeAt(i));
    }

  const result = process.end(str);

  return result;
}

// based on cyrb53
export function getHashProcessor(seed = 0) {
  let h1 = 0xdeadbeef ^ seed;
  let h2 = 0x41c6ce57 ^ seed;

  return {
    next(_str: string, ch: number) {
      h1 = Math.imul(h1 ^ ch, 2654435761);
      h2 = Math.imul(h2 ^ ch, 1597334677);
    },
    end() {
      h1 =
        Math.imul(h1 ^ (h1 >>> 16), 2246822507) ^
        Math.imul(h2 ^ (h2 >>> 13), 3266489909);
      h2 =
        Math.imul(h2 ^ (h2 >>> 16), 2246822507) ^
        Math.imul(h1 ^ (h1 >>> 13), 3266489909);

      return (
        (h2 >>> 0).toString(16).padStart(8, "0") +
        (h1 >>> 0).toString(16).padStart(8, "0")
      );
    },
  };
}

type CSSSelectorTree = {
  selector: string;
  rules: (string | StyleTree)[];
};
type StyleTree =
  | CSSSelectorTree
  | {
      wrapper: string;
      subTrees: StyleTree[];
    };

export function getUnnestProcessor() {
  let tree: StyleTree = {
    selector: "",
    rules: [""],
  };
  let buffer = "";
  let quote = "";
  let isWrapping = false;
  let isWhitespace = false;
  let cursors: StyleTree[] = [tree];
  const appendRule = (str: string) => {
    let cursor = cursors[cursors.length - 1];
    if (!("rules" in cursor)) throw new Error();
    if (typeof cursor.rules[cursor.rules.length - 1] !== "string")
      cursor.rules.push(str);
    else cursor.rules[cursor.rules.length - 1] += str;
  };

  return {
    next(str: string, _char: number) {
      let cursor = cursors[cursors.length - 1];

      if ((str === '"' || str === "'") && !quote) {
        buffer += str;
        quote = str;
      } else if (quote === str) {
        buffer += str;
        quote = "";
      } else if (quote) {
        buffer += str;
      } else if (str === ";" && "rules" in cursor) {
        appendRule(buffer.trim() + ";");
        buffer = "";
      } else if (str === "@" && buffer.trim().length === 0) {
        buffer = "@";
        isWrapping = true;
      } else if (str === "{" && isWrapping) {
        let newTree: StyleTree = {
          wrapper: buffer.trim(),
          subTrees: [
            {
              selector: "&",
              rules: [""],
            },
          ],
        };
        cursors.push(newTree.subTrees[0]);
        (cursor as CSSSelectorTree).rules.push(newTree);
        buffer = "";
        isWrapping = false;
      } else if (str === "{") {
        let newTree: StyleTree = {
          selector: buffer.trim(),
          rules: [""],
        };

        cursors.push(newTree);
        (cursor as CSSSelectorTree).rules.push(newTree);

        buffer = "";
      } else if (str === "}") {
        if ("rules" in cursor) {
          appendRule(buffer);
          buffer = "";
        }
        cursors.pop();
      } else if (" \t\n".includes(str)) {
        if (isWhitespace) return;
        isWhitespace = true;
        buffer += " ";
      } else {
        isWhitespace = false;
        buffer += str;
      }
    },
    end() {
      let cursor = cursors[cursors.length - 1];

      if ("rules" in cursor) {
        appendRule(buffer);
        buffer = "";
      }

      function walk(tree: StyleTree, topSelector: string): string {
        if ("wrapper" in tree) {
          return `${tree.wrapper}{${tree.subTrees
            .map((tree) => walk(tree, topSelector))
            .filter((x) => x)
            .join("\n")}}`;
        }

        if (topSelector === "&") topSelector = "";

        const { selector, rules } = tree;
        const newSelector = !topSelector
          ? selector
          : topSelector
              .split(",")
              .map((s) => s.trim())
              .map((topSelector) =>
                selector
                  .split(",")
                  .map((s) => s.trim())
                  .map((selector) =>
                    selector.includes("&")
                      ? selector.replace(/&/g, topSelector)
                      : `${topSelector} ${selector}`
                  )
                  .join(",")
              )
              .join(",")
              .trim();

        let output = [];
        for (let rule of rules) {
          if (typeof rule === "string") {
            const ruleText = rule.trim();
            const selector = newSelector.trim();
            if (ruleText && selector) output.push(`${selector}{${ruleText}}`);
          } else {
            const result = walk(rule, newSelector);
            if (result) output.push(result);
          }
        }

        return output.join("\n");
      }

      return (topLevel: string) => walk(tree, topLevel);
    },
  };
}
