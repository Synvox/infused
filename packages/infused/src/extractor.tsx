import { ReactNode } from "react";
import ReplaceStream from "replacestream";
import { Stream } from "stream";
import { CSSExtractionProvider } from "./context";
import { InfusedStyleSheet } from "./stylesheet";

export function getExtractor(
  replacedString: string = "ðŸŽ¨_INTERNAL_STYLES_DO_NOT_USE_ðŸŽ¨"
) {
  const sheets: Set<InfusedStyleSheet> = new Set();

  function Provider({ children }: { children: ReactNode }) {
    return (
      <CSSExtractionProvider
        children={children}
        sheets={sheets}
        replacedString={replacedString}
      />
    );
  }

  function addStyles(markup: string): string;
  function addStyles(markup: Stream): Stream;
  function addStyles(markup: string | Stream): string | Stream {
    const sortedSheets = Array.from(sheets).sort((a, b) => a.index - b.index);

    const styleTags = sortedSheets
      .map(({ id, css }) => `<style id="${id}">${css}</style>`)
      .join("");

    if (typeof markup === "string")
      return markup.replace(replacedString, styleTags);
    else return markup.pipe(ReplaceStream(replacedString, styleTags));
  }

  return { Provider, addStyles };
}
