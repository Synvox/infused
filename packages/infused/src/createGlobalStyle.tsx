import { FunctionComponent } from "react";
import { getUnnestProcessor } from "./processor";
import { ValidInterpolation } from "./styled";
import { InfusedStyleSheet } from "./stylesheet";
import { template } from "./template";
import { useTheme } from "./theme";

export interface GlobalStyle {
  <Props>(
    s: TemplateStringsArray,
    ...p: ValidInterpolation<Props>[]
  ): FunctionComponent<Props>;
}

export function _createGlobalStyle(stylesheet: InfusedStyleSheet): GlobalStyle {
  return template(
    stylesheet,
    (_id, useStyle, parseProps) => {
      return function (props: any) {
        useStyle();
        const theme = useTheme();
        const parsed = Object.entries(parseProps({ ...props, theme }))
          .map(([k, v]) => `${k}:${v}`)
          .join(";");

        if (!parsed) return null;
        return (
          <style dangerouslySetInnerHTML={{ __html: `:root{${parsed}}` }} />
        );
      };
    },
    () => {
      const p = getUnnestProcessor();
      return {
        next: p.next,
        end() {
          return () => p.end()("");
        },
      };
    }
  );
}
