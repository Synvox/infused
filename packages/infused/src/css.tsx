import { ThemeProp, ValidInterpolation } from "./styled";
import { InfusedStyleSheet } from "./stylesheet";
import { parseTemplateString } from "./template";

export class CSSFragment<P> {
  string: string;
  dependsOn: Set<InfusedStyleSheet>;
  fns: ((props: P & ThemeProp) => string | CSSFragment<P>)[];
  constructor(p: {
    string: typeof CSSFragment.prototype.string;
    dependsOn: typeof CSSFragment.prototype.dependsOn;
    fns: typeof CSSFragment.prototype.fns;
  }) {
    this.string = p.string;
    this.dependsOn = p.dependsOn;
    this.fns = p.fns;
  }
}

export function _css(stylesheet: InfusedStyleSheet) {
  return function css<P>(
    strings: TemplateStringsArray,
    ...args: ValidInterpolation<P>[]
  ): CSSFragment<P> {
    const { string, dependsOn, fns } = parseTemplateString(
      stylesheet,
      strings,
      ...args
    );

    return new CSSFragment({ string, dependsOn, fns });
  };
}
