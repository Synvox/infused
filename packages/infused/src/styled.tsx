import isPropValid from "@emotion/is-prop-valid";
import {
  ComponentType,
  createElement,
  forwardRef,
  FunctionComponent,
  PropsWithoutRef,
  RefAttributes,
} from "react";
import { DefaultTheme } from ".";
import { CSSFragment } from "./css";
import { CSSKeyframes } from "./keyframes";
import { getUnnestProcessor } from "./processor";
import { InfusedStyleSheet, WithInfusedStyleSheet } from "./stylesheet";
import { template } from "./template";
import { useTheme } from "./theme";

type HTMLTagName = keyof JSX.IntrinsicElements;

export class StyledMeta extends WithInfusedStyleSheet {
  className: string;
  type: string;
  constructor(p: {
    className: typeof StyledMeta.prototype.className;
    type: typeof StyledMeta.prototype.type;
    stylesheet: typeof StyledMeta.prototype.stylesheet;
  }) {
    super(p.stylesheet);
    this.className = p.className;
    this.type = p.type;
  }
}

interface StyledBuilder<BaseProps> {
  <Props>(
    s: TemplateStringsArray,
    ...p: ValidInterpolation<BaseProps & Props>[]
  ): StyledComponent<BaseProps & Props>;
  classes: <Props>(
    classes: Record<string, (p: BaseProps & Props) => any>
  ) => StyledBuilder<Props>;
  attrs: <Props>(
    attributes: Record<string, any>
  ) => StyledBuilder<BaseProps & Props>;
}

export interface StyledComponent<Props>
  extends FunctionComponent<Props & { as?: string | ComponentType }> {
  __styled: StyledMeta;
  classes: (
    classes: Record<string, (p: Props) => any>
  ) => StyledComponent<Props>;
  attrs: (attributes: Record<string, any>) => StyledComponent<Props>;
}

export interface FunctionInterpolation<P> {}

export type ValidInterpolation<P> =
  | { __styled: StyledMeta }
  | CSSKeyframes
  | CSSFragment<P & ThemeProp>
  | (<Props extends P = P>(
      p: Props & ThemeProp
    ) => string | number | undefined | null | false)
  | string
  | number
  | undefined
  | null
  | false;

export type ThemeProp = {
  theme: DefaultTheme;
};

type StyledProxy = {
  [Tag in HTMLTagName]: StyledBuilder<JSX.IntrinsicElements[Tag]>;
};

export interface Styled extends StyledProxy {
  <P, T extends HTMLTagName>(tag: T): StyledBuilder<
    JSX.IntrinsicElements[T] & P
  >;
  <P, T = P extends RefAttributes<infer X> ? X : never>(
    tag: FunctionComponent<P>
  ): StyledBuilder<PropsWithoutRef<P> & RefAttributes<T>>;
  <P>(tag: ComponentType<P>): StyledBuilder<P>;
  <P>(tag: StyledComponent<P>): StyledBuilder<P>;
}

export function _styled(stylesheet: InfusedStyleSheet): Styled {
  function makeStyledComponent(givenType: any) {
    const metaData: StyledMeta =
      givenType.styled instanceof StyledMeta ? givenType.styled : undefined;
    const type = metaData ? metaData.type : givenType;

    function buildFunctionComponent(
      className: string,
      useStyle: () => void,
      parseProps: (p: any) => any,
      {
        classes = {},
        attrs = {},
      }: {
        classes?: Record<string, (p: any) => any>;
        attrs?: Record<string, any>;
      } = {}
    ) {
      return Object.assign(
        forwardRef(function Component(props: any, ref) {
          useStyle();
          const theme = useTheme();

          const classNames = [];
          for (let className of Object.keys(classes)) {
            if (classes[className]({ ...props, theme }))
              classNames.push(className);
          }

          classNames.unshift(className);
          if (metaData) classNames.unshift(metaData.className);
          if (props.className) classNames.unshift(props.className);

          const cleanedProps =
            typeof type === "string"
              ? Object.fromEntries(
                  Object.entries(props).filter(([attr]) => isPropValid(attr))
                )
              : props;

          return createElement(props.as ? props.as : type, {
            ...attrs,
            ...cleanedProps,
            style: {
              ...parseProps({ ...props, theme }),
              ...props.style,
            },
            className: classNames.join(" "),
            ref,
          });
        }),
        {
          classes(c: Record<string, (props: object) => boolean>) {
            return buildFunctionComponent(className, useStyle, parseProps, {
              attrs,
              classes: {
                ...classes,
                ...c,
              },
            });
          },
          attrs(a: Record<string, (props: object) => boolean>) {
            return buildFunctionComponent(className, useStyle, parseProps, {
              classes,
              attrs: {
                ...attrs,
                ...a,
              },
            });
          },
          __styled: new StyledMeta({
            className: metaData
              ? [metaData.className, className].join(" ")
              : className,
            type: givenType,
            stylesheet,
          }),
        }
      );
    }

    function makeTemplate({
      classes = {},
      attrs = {},
    }: {
      classes?: Record<string, (p: any) => any>;
      attrs?: Record<string, any>;
    } = {}) {
      return Object.assign(
        template(
          stylesheet,
          (className, useStyle, parseProps) => {
            return buildFunctionComponent(className, useStyle, parseProps, {
              classes,
              attrs,
            });
          },
          () => {
            const p = getUnnestProcessor();
            return {
              next: p.next,
              end() {
                return (id) => p.end()(`.${id}`);
              },
            };
          }
        ),
        {
          classes(c: Record<string, (props: object) => boolean>) {
            return makeTemplate({ attrs, classes: { ...classes, ...c } });
          },
          attrs(a: Record<string, (props: object) => boolean>) {
            return makeTemplate({ classes, attrs: { ...attrs, ...a } });
          },
        }
      );
    }

    return makeTemplate();
  }

  //@ts-expect-error
  return new Proxy(makeStyledComponent, {
    get(_, tagName, fn) {
      return fn(tagName);
    },
  });
}
