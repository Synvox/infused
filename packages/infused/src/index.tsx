import {
  GlobalStyle,
  _createGlobalStyle as createGlobalStyle,
} from "./createGlobalStyle";
import { CSSFragment, _css as css } from "./css";
import { CSSKeyframes, _keyframes as keyframes } from "./keyframes";
import { StyledComponent, _styled as styled } from "./styled";
import { InfusedStyleSheet } from "./stylesheet";
export { CSSExtractionProvider, Styles } from "./context";
export { getExtractor } from "./extractor";
export { ThemeProvider, useTheme } from "./theme";
export { InfusedStyleSheet };
export type { GlobalStyle, CSSFragment, CSSKeyframes, StyledComponent };

export function StyleSheet(prefix?: string) {
  const s = new InfusedStyleSheet(prefix);

  return {
    styled: styled(s),
    createGlobalStyle: createGlobalStyle(s),
    keyframes: keyframes(s),
    css: css(s),
    toString() {
      s.compile();
      return s.css;
    },
  };
}

export interface DefaultTheme {}
