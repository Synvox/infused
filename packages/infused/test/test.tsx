import { it } from "@jest/globals";
import { StyleSheet } from "../src";

it("compiles simple tags", () => {
  const { styled, toString } = StyleSheet();

  styled.div`
    background-color: blue;
  `;

  expect(toString()).toMatchSnapshot();
});

it("compiles interpolations", () => {
  const { styled, toString } = StyleSheet();

  styled.div`
    background-color: ${"blue"};
  `;

  expect(toString()).toMatchSnapshot();
});

it("compiles function interpolations", () => {
  const { styled, toString } = StyleSheet();

  styled.div`
    background-color: ${() => "blue"};
  `;

  expect(toString()).toMatchSnapshot();
});

it("compiles css interpolations", () => {
  const { styled, toString, css } = StyleSheet();

  const shared = css`
    margin: auto;
  `;

  styled.div`
    ${shared}
    background-color: ${() => "blue"};
  `;

  expect(toString()).toMatchSnapshot();
});

it("compiles keyframes interpolations", () => {
  const { styled, toString, keyframes } = StyleSheet();

  const anim = keyframes`
    0% {
      color:black;
    }
    100% {
      color:white;
    }
  `;

  styled.div`
    animation: 1s ${anim};
  `;

  expect(toString()).toMatchSnapshot();
});

it("compiles nested rules", () => {
  const { styled, toString } = StyleSheet();

  styled.div`
    color: black;
    &:hover {
      color: white;
      a {
        text-decoration: dashed;
        @media (max-width: 500px) {
          display: block;
        }
      }
    }
  `;

  expect(toString()).toMatchSnapshot();
});

it("compiles global styles", () => {
  const { createGlobalStyle, toString } = StyleSheet();

  createGlobalStyle`body {margin: 0;}`;

  expect(toString()).toMatchSnapshot();
});
