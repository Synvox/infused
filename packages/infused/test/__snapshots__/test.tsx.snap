// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adds 1 + 2 to equal 3 1`] = `".i698f458c944cbfdb{background-color: blue;}"`;

exports[`compile function interpolations 1`] = `".ibb4b2b09abc56981{background-color: var(--ibb4b2b09abc56981-0);}"`;

exports[`compiles css interpolations 1`] = `".if0d56cc259a2964a{margin: auto;background-color: var(--if0d56cc259a2964a-0);}"`;

exports[`compiles function interpolations 1`] = `".ibb4b2b09abc56981{background-color: var(--ibb4b2b09abc56981-0);}"`;

exports[`compiles global styles 1`] = `"body{margin: 0;}"`;

exports[`compiles interpolations 1`] = `".i698f458c944cbfdb{background-color: blue;}"`;

exports[`compiles keyframes interpolations 1`] = `
"@keyframes i9e23bee7ec1724d2{ 0% { color:black; } 100% { color:white; } }
.ia426940a7b5a2b3c{animation: 1s i9e23bee7ec1724d2;}"
`;

exports[`compiles nested rules 1`] = `
".ie140a6064e781688{color: black;}
.ie140a6064e781688:hover{color: white;}
.ie140a6064e781688:hover a{text-decoration: dashed;}
@media (max-width: 500px){.ie140a6064e781688:hover a{display: block;}}"
`;

exports[`compiles simple tags 1`] = `".i698f458c944cbfdb{background-color: blue;}"`;
